etna/transforms/missing_values/imputation.py

```diff
@@ -150,24 +185,20 @@ def _transform(self, df: pd.DataFrame) -> pd.DataFrame:
         :
             Transformed Dataframe in etna wide format.
         """
-        if self._fill_value is None or self._nan_timestamps is None:
+        if self._fit_segments is None or self._nans_to_impute_mask is None:
             raise ValueError("Transform is not fitted!")
 
-        segments = sorted(set(df.columns.get_level_values("segment")))
-        check_new_segments(transform_segments=segments, fit_segments=self._nan_timestamps.keys())
+        segments = df.columns.get_level_values("segment").unique()
+        check_new_segments(transform_segments=segments, fit_segments=self._fit_segments)
 
-        cur_nans = {}
-        for segment in segments:
-            series = df.loc[:, pd.IndexSlice[segment, self.in_column]]
-            cur_nans[segment] = series[series.isna()].index
+        index_intersection = df.index.intersection(self._nans_to_impute_mask.index)
+        nans_to_restore_mask = df.isna()
+        nans_to_restore_mask.loc[index_intersection] = nans_to_restore_mask.loc[index_intersection].mask(
+            self._nans_to_impute_mask.loc[index_intersection], False
```

Comment:
I don't really understand what is happening here.

Will it work fine if `self._nans_to_impute_mask` has more segments than `df`.

Reply:
Here we mark nans that should be imputed(from `self. _nans_to_impute_mask`) with `False` as they shouldn't be restored after imputation